Data/Vinyl/Curry.hs

```diff
@@ -0,0 +1,125 @@
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+
+{-|
+
+Provides combinators for currying and uncurrying functions over arbitrary vinyl
+records.
+
+-}
+module Data.Vinyl.Curry where
+
+import           Data.Vinyl
+import           Data.Vinyl.Functor
+
+-- * Currying
+
+class RecordCurry ts where
+  -- | N-ary version of 'curry' over functorial records.
+  rcurry :: (Rec f ts -> a) -> CurriedF f ts a
+
+  -- | N-ary version of 'curry' over pure records.
+  rcurry' :: (Rec Identity ts -> a) -> Curried ts a
+
+
+instance RecordCurry '[] where
+  rcurry f = f RNil
+  {-# INLINABLE rcurry #-}
+  rcurry' f = f RNil
+  {-# INLINABLE rcurry' #-}
+
+instance RecordCurry ts => RecordCurry (t ': ts) where
+  rcurry f x = rcurry (\xs -> f (x :& xs))
+  {-# INLINABLE rcurry #-}
+  rcurry' f x = rcurry' (\xs -> f (Identity x :& xs))
+  {-# INLINABLE rcurry' #-}
+
+-- * Uncurrying
+
+-- | N-ary version of 'uncurry' over functorial records.
```

Comment:
Some examples of the specialized type of `rcurry` and `runcurry` on small records would help those unfamiliar with type families.

> runcurry :: (f x -> f y -> a) -> Rec f '[x, y] -> a
> -- etc.